clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java

```diff
@@ -441,4 +535,43 @@ private Result checkDistributors(Node node, int clusterStateVersion) {
         return allowSettingOfWantedState();
     }
 
+    private Set<Integer> groupsWithUserWantedStateNotUp() {
+        return clusterInfo.getAllNodeInfos().stream()
+                          .filter(sni -> !UP.equals(sni.getUserWantedState().getState()))
+                          .map(NodeInfo::getGroup)
+                          .filter(Objects::nonNull)
+                          .filter(Group::isLeafGroup)
+                          .map(Group::getIndex)
+                          .collect(Collectors.toSet());
+    }
+
+    // groups with at least one node with the same state & description
+    private Set<Integer> groupsWithSameStateAndDescription(State state, String newDescription) {
+        return clusterInfo.getAllNodeInfos().stream()
+                          .filter(nodeInfo -> {
+                              var userWantedState = nodeInfo.getUserWantedState();
+                              return userWantedState.getState() == state &&
+                                      Objects.equals(userWantedState.getDescription(), newDescription);
+                          })
+                          .map(NodeInfo::getGroup)
+                          .filter(Objects::nonNull)
+                          .filter(Group::isLeafGroup)
+                          .map(Group::getIndex)
+                          .collect(Collectors.toSet());
+    }
+
+    // groups with at least one node with the same state & description
+    private Set<Integer> retiredOrNotUpGroups(ClusterState clusterState) {
+        return clusterInfo.getAllNodeInfos().stream()
+                          .filter(nodeInfo -> (nodeInfo.getUserWantedState().getState() == RETIRED
```

Comment:
Filter those that have user wanted state *!=* RETIRED *and* != UP, *or* those that have state *!=* RETIRED *and* != UP.  We essentially treat RETIRED as UP in these checks.  If either the user wanted state or the state is not up/retired, we want to add the corresponding group to the returned set.

Reply:
Fixed